Make TS recognize input arg name

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73253

This PR allows TS schema_matching to match input arg with self for aten operators. This is because, operators in their functional form have input as paremeter instead of self. 

fixes: https://github.com/pytorch/pytorch/issues/71994

Differential Revision: [D34427556](https://our.internmc.facebook.com/intern/diff/D34427556)

tugsbayasgalan committed:
Make TS recognize input arg name

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make TS recognize input arg name"


fixes: https://github.com/pytorch/pytorch/issues/71994

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/frontend/schema_matching.cpp:
```diff
@@ -230,10 +230,17 @@ static Value* tryMatchArgument(
     const std::string& name,
-    at::ArrayRef<NamedValue> kwargs) {
+    at::ArrayRef<NamedValue> kwargs,
+    bool is_aten) {
```
Is it necessary to check `is_aten`? What if we simply allow `input` to be an alias for `self` universally? (unless there is another argument that's called input). What do you think?

eellison replied:
if there's a user defined function with both `input` and `self` defined something weird could happen

anjali411 replied:
```
def f(x):
            return torch.ceil(self=x)

        inp = torch.randn(10)
        self.checkScript(f, (inp, ))
```

doesn't work on master, so should be ok

tugsbayasgalan replied:
We can have operators with input as their parameter name in the future. So it is probably better to do it for only aten ops for now? @gmagogsfm

gmagogsfm replied:
Yeah, we should only allow this aliasing when there is no other argument with `self`, `input` as their names. That should be safe.

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34427556).

gmagogsfm reviewed (approved):
Are there any other cases where the schema defined in `native_functions.yaml` is changed during code generation? Because we'd need to copy that change in TorchScript too. Otherwise, it would always treat the original (unchanged) schema in `native_functions.yaml` as source of truth. @anjali411

tugsbayasgalan committed:
Update on "Make TS recognize input arg name"


This PR allows TS schema_matching to match input arg with self for aten operators. This is because, operators in their functional form have input as paremeter instead of self. 

fixes: https://github.com/pytorch/pytorch/issues/71994

Differential Revision: [D34427556](https://our.internmc.facebook.com/intern/diff/D34427556)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34427556).

anjali411 commented:
> Are there any other cases where the schema defined in `native_functions.yaml` is changed during code generation? Because we'd need to copy that change in TorchScript too. Otherwise, it would always treat the original (unchanged) schema in `native_functions.yaml` as source of truth. @anjali411

I am not sure tbh
